sw/device/lib/dif/dif_gpio.c

```diff
@@ -0,0 +1,169 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#include "sw/device/lib/dif/dif_gpio.h"
+
+#include "gpio_regs.h"  // Generated.
+#include "sw/device/lib/common.h"
+
+// This is a temporary measure to work around the dependency on
+// |GPIO0_BASE_ADDR| in |GPIO_*()| family of macros in gpio_regs.h.
+// See lowrisc/opentitan#1118.
+static const uintptr_t GPIO0_BASE_ADDR = 0x0U;
+#define DEFINE_GPIO_REG_CONSTANT(NAME) \
+  static const ptrdiff_t GPIO_##NAME##_REG_OFFSET = GPIO_##NAME(0);
+DEFINE_GPIO_REG_CONSTANT(INTR_STATE)
+DEFINE_GPIO_REG_CONSTANT(INTR_ENABLE)
+DEFINE_GPIO_REG_CONSTANT(INTR_TEST)
+DEFINE_GPIO_REG_CONSTANT(DATA_IN)
+DEFINE_GPIO_REG_CONSTANT(DIRECT_OUT)
+DEFINE_GPIO_REG_CONSTANT(MASKED_OUT_LOWER)
+DEFINE_GPIO_REG_CONSTANT(MASKED_OUT_UPPER)
+DEFINE_GPIO_REG_CONSTANT(DIRECT_OE)
+DEFINE_GPIO_REG_CONSTANT(MASKED_OE_LOWER)
+DEFINE_GPIO_REG_CONSTANT(MASKED_OE_UPPER)
+DEFINE_GPIO_REG_CONSTANT(INTR_CTRL_EN_RISING)
+DEFINE_GPIO_REG_CONSTANT(INTR_CTRL_EN_FALLING)
+DEFINE_GPIO_REG_CONSTANT(INTR_CTRL_EN_LVLHIGH)
+DEFINE_GPIO_REG_CONSTANT(INTR_CTRL_EN_LVLLOW)
+DEFINE_GPIO_REG_CONSTANT(CTRL_EN_INPUT_FILTER)
+#undef DEFINE_GPIO_REG_CONSTANT
+
+/**
+ * Gives the mask that corresponds to the given bit index.
+ *
+ * @param index Bit index in a 32-bit register.
+ */
+static inline uint32_t index_to_mask(const uint32_t index) {
+  return 1U << index;
+}
+
+/**
+ * Perform a masked write to a register.
+ *
+ * GPIO device provides masked (thread-safe) writes to some of the registers.
+ * This allows software to modify half of the bits at a time without requiring
+ * a read-modify-write. Note that depending on the value of the |mask|, this
+ * function may perform two writes.
+ *
+ * @param gpio GPIO instance
+ * @param lowerOffset Offset of the masked access register to modify the lower
+ * half of a register.
+ * @param upperOffset Offset of the masked access register to modify the upper
+ * half of a register.
+ * @param mask Mask that identifies the bits to write to.
+ * @param val Value to write
+ */
+static inline void write_register_masked_access(const dif_gpio_t *const gpio,
+                                                const uint32_t lowerOffset,
+                                                const uint32_t upperOffset,
+                                                const uint32_t mask,
+                                                const uint32_t val) {
+  const uint32_t lower_mask = mask & 0x0000FFFFU;
+  const uint32_t upper_mask = mask & 0xFFFF0000U;
+  if (lower_mask != 0) {
+    reg32_write(gpio->base_addr, lowerOffset,
+                (lower_mask << 16) | (val & 0x0000FFFFU));
+  }
+  if (upper_mask != 0) {
+    reg32_write(gpio->base_addr, upperOffset,
+                upper_mask | ((val & 0xFFFF0000U) >> 16));
+  }
+}
+
+static inline void write_register_bit_masked_access(
```

Comment:
Please consider splitting this function in two:
```
void gpio_set_masked_upper_bit(const dif_gpio_t *const gpio, const uint32_t index, const bool val);
void gpio_set_masked_lower_bit(const dif_gpio_t *const gpio, const uint32_t index, const bool val);
```
You probably could even implement these functions in terms of:
```
void gpio_set_masked_upper(const dif_gpio_t *const gpio, const uint32_t mask, const uint32_t val)
void gpio_set_masked_lower(const dif_gpio_t *const gpio, const uint32_t mask, const uint32_t val)
```

Reply:
I'm not sure if this would be helpful because we will still need the logic to determine which half we should modify as well as some bit operations to shift the mask bit to the correct positions. The rest would be just calls to `reg32_write`. That being said, I will leave this open if you have further comments or if I need to revisit this.

Reply:
I was thinking of something like this:
```
...
#define GPIO_32BIT_REG_UPPER_HALF_MASK 0xffff0000u
#define GPIO_32BIT_REG_LOWER_HALF_MASK 0x0000ffffu
#define GPIO_UPPER_16BIT_OFFSET 16u

static void gpio_masked_write_upper(const dif_gpio_t *gpio,
                                    ptrdiff_t register_offset,
                                    uint16_t enable_mask, uint16_t value) {
  uint32_t register_value = ((uint32_t)enable_mask << GPIO_UPPER_16BIT_OFFSET) |
                            ((uint32_t)value & GPIO_32BIT_REG_LOWER_HALF_MASK);

  mmio_region_write32(gpio->base_addr, register_offset, register_value);
}

static void gpio_masked_write_lower(const dif_gpio_t *gpio,
                                    ptrdiff_t register_offset,
                                    uint16_t enable_mask, uint16_t value) {
...
}

bool dif_gpio_masked_out_write_upper(const dif_gpio_t *gpio,
                                     dif_gpio_masked_op_type type,
                                     uint16_t mask, uint16_t val) {
  ptrdiff_t offset = GPIO_MASKED_OUT_UPPER_REG_OFFSET;
  switch(type) {
  case kDifGpioOut:
    offset = GPIO_MASKED_OUT_UPPER_REG_OFFSET;
    break;
  ...
  default:
    return false;
  }

  gpio_masked_write_upper(gpio, GPIO_MASKED_OUT_UPPER_REG_OFFSET, mask, val);

  return true;
}

bool dif_gpio_masked_out_write_lower(const dif_gpio_t *gpio,
                                     dif_gpio_masked_op_type type,
                                     uint16_t mask, uint16_t val) {
  ...
}
...
```

Reply:
I think this allows you to implement things like `gpio_masked_bit_write_upper/lower` in terms of `gpio_masked_write_upper/lower` easily.

Reply:
I see, thank you for the explanation and the code! I will leave this conversation open while I think about this.